numpy/core/src/npysort/selection.c.src

```diff
@@ -406,6 +448,107 @@ NPY_NO_EXPORT int
 }
 
 
+/*
+ * rivest-floyd selection algorithm
+ */
+#define C1 600
+#define C2 0.5
+#define C3 0.5
+
+/* helper function */
+void
+@name@rf_recursive_@suff@(@type@ *v,
+#if @arg@
+                          npy_intp* tosort,
+#endif
+                          npy_intp left,
+                          npy_intp right,
+                          npy_intp kth)
+{
+    npy_intp j;
+    while (right > left) {
+        if (right - left > C1) {
+            npy_float N = (npy_float)(right - left + 1);
+            npy_float II = (npy_float)(kth - left + 1);
+            npy_float Z = npy_log(N);
+            npy_float S = C2 * npy_exp(2*Z/3.0);
+            npy_float SD = C3 * npy_sqrt(Z*S*(N - S)/N) * npy_copysign(1, II - N/2);
+            npy_intp newleft = kth - (npy_intp)(II*S/N - SD);
+            /* why II*S/N - SD can be converted to npy_intp and subtracted from kth without numerical overflow:
+            1) converted to npy_intp
+            Since N >= 600,
+            0 <= II <= N,
+            S = 0.5 N^(2/3) and
+            |SD| = 0.5 sqrt(log N * 0.5 N^(2/3) * (N - 0.5 N^(2/3) / N) <= 0.5 N^(1/3) sqrt(log N):
+            |II*S/N - SD| <= |II*S/N| + |SD| <= |S| + |SD| < N <= npy_intp.max
```

Comment:
Could you reformat this as block comments starting with ` * ` on each line?
```
/*
 * This is a
 * long comment
 */
```
Technically, it appears that the last `N <= npy_intp.max` is not necessary correct, but it does not matter as long as the previous `|S| + |SD| < N` is strictly right (see below).

Anyway, I don't think this matters, just trying it out, S and SD are much smaller numbers than the original input (and I expect that has to be the case, since this recursively brackets the result. As for small numbers ~600, there isn't any issue to begin with.

Could you apply all the style suggestions, I expect this should be pretty much ready go (but didn't look at everything yet).

<details>
```
In [10]: np.int32((np.float32(np.iinfo(np.int32).max)))
Out[10]: -2147483648

In [11]: np.int32(np.nextafter(np.float32(np.iinfo(np.int32).max), 0))
Out[11]: 2147483647
```

Reply:
Thank you! I've applied @eric-wieser's and your style suggestions, and thanks, @madphysicist ,  I wondered how to name it because floydrivestselect seemed too long, but floydrivest is great, I changed it.

Reply:
My gut feeling is dropping the "select", unlike with "introselect", where the "select" part actually distinguishes it from introsort, I think the "select" has no actual meaning here. We could also do `floyd-rivest` quite frankly (in python).

Reply:
Oops forgot to change the documentation in one place, changed now.